File: Napakalaki Ruby/Player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'BadConsequence.rb'
require_relative 'NumericBadConsequence.rb'
require_relative 'DeathBadConsequence.rb'
require_relative 'SpecificBadConsequence.rb'
require_relative 'Monster.rb'
require_relative 'TreasureKind.rb'
require_relative 'Treasure.rb'
require_relative 'Dice.rb'
require_relative 'CombatResult.rb'

module NapakalakiGame

class Player
  
  attr_reader :name, :level, :dead, :can_i_steal, :hidden_treasures, :visible_treasures, :combat_level
  attr_accessor :pending_bad_consequence, :enemy
  
  @@MAXLEVEL = 10
  
  def initialize(nombre)
    
    @name = nombre
    @level = 0
    @dead = true
    @can_i_steal = true
    @hidden_treasures = Array.new
    @visible_treasures = Array.new
    @combat_level = 0
    @pending_bad_consequence = NumericBadConsequence.new(nil,0,0,0)
    @enemy = nil
    
  end
   
  private
  
  def bringToLife()
    @dead = false
  end
  
  def incrementLevels(i)
    @level = level + i
  end
  
  
  def decrementLevels(i)
    if ((@level - i) < 0)
      @level = 0
    else 
      @level = @level - i
    end
  end
  
  
  def setPendingBadConsequence(b) #BadConsequence b
    @pending_bad_consequence = b
  end
  
  
  def applyPrize(m) #Monster m
    n_levels = m.getLevelGained
    incrementLevels(n_levels)
    n_treasures = m.getTreasuresGained
    if (n_treasures > 0)
      dealer = CardDealer.instance
      i=0
      while i < n_treasures do
        treasure = dealer.nextTreasure
        @hidden_treasures << treasure
        i+=1
      end
    end
  end
  

  def applyBadConsequence(m) #Monster m
    bad_consequence = m.getBadConsequence
    n_levels = bad_consequence.getLevels
    decrementLevels(n_levels)
    pending_bad = bad_consequence.adjustToFitTreasureList(@visible_treasures,@hidden_treasures)
    setPendingBadConsequence(pending_bad)
  end
  
  
  def canMakeTreasureVisible(t) #Treasure t
      can_make = false
      #Comprobamos si tiene otra armadura,casco o botas si la tiene, no puede hacerse visible
      if (t.getType == TreasureKind::ARMOR or 
          t.getType == TreasureKind::HELMET or 
          t.getType == TreasureKind::SHOES)
          can_make = (howManyVisibleTreasures(t.getType) == 0)
      #Comprobamos si tiene otro arma a dos manos o un arma a una mano, si la tiene, no puede hacerse visible
      #si tiene aguno de los dos, no puede hacerse visible
      else if (t.getType == TreasureKind::BOTHHANDS)
          can_make = (howManyVisibleTreasures(TreasureKind::BOTHHANDS) == 0 and 
                      howManyVisibleTreasures(TreasureKind::ONEHAND)== 0)
      #Comprobamos si tiene un arma a dos manos o mas de un arma a una mano,
      #si es asi, no puede hacerse visible
      else
          can_make = (howManyVisibleTreasures(TreasureKind::BOTHHANDS) == 0 and 
                      howManyVisibleTreasures(TreasureKind::ONEHAND) < 2)
      end end

    return can_make
  end
  
  
  def howManyVisibleTreasures(t_kind)
    n_treasures = 0
    
    for i in @visible_treasures
      n_treasures += 1 if (i.getType == t_kind)
    end
    
    return n_treasures
  end
  
  def dieIfNoTreasures()
     
    if (@hidden_treasures.size == 0 and @visible_treasures.size == 0)
    @dead = true 
    @level = 1
    end
  end
  
  def haveStolen()
    @can_i_steal = true
  end
  
  protected
  
  def getOponentLevel(m)
    n = m.getCombatLevel
    return n
  end
  
  def shouldConvert()
    dado = Dice.instance
    n = dado.nextNumber
    if(n == 1)
      aux = true
    else
      aux = false
    end
    
    return aux
  end
  
  def getCombatLevel()
    nivel = @level  
    
    @visible_treasures.each do |tesoro|
      nivel += tesoro.getBonus
    end
    
    return nivel; 
  end
  
  public
  
  def copiarJugador(p_aux)
    @name = p_aux.getName
    @level = p_aux.getLevels
    @dead = p_aux.isDead
    @can_i_steal = p_aux.canISteal
    @hidden_treasures =p_aux.getHiddenTreasures
    @visible_treasures = p_aux.getVisibleTreasures
    @combat_level = p_aux.getCombatLevel
    @pending_bad_consequence = p_aux.pending_bad_consequence
    @enemy = p_aux.enemy
  end
  
  def to_s()
    return @name
  end
  
  def getName() 
    return @name
  end
   

  def getLevels()
    return @level
  end
  
  def isDead()
    return @dead
  end
  
  
  def combat(monstruo)
    my_level = getCombatLevel
    monster_level = getOponentLevel(monstruo)

    if (my_level > monster_level)
      applyPrize(monstruo)
      combat_result = CombatResult::WIN
    else
      applyBadConsequence(monstruo)
      combat_result = CombatResult::LOSE
      if(shouldConvert)
        combat_result = CombatResult::LOSEANDCONVERT
      end
    end
    
    if(@level > 9)
      combat_result = CombatResult::WINGAME
    end
    
    return combat_result
  end
  
  
  def makeTreasureVisible(tesoro)
    can_i = canMakeTreasureVisible(tesoro)
    if (can_i)
      @visible_treasures << tesoro
      @hidden_treasures.delete(tesoro)
    end
  end
  
  
  def discardVisibleTreasure(tesoro)
    @visible_treasures.delete(tesoro)
    if(@pending_bad_consequence!=nil and (!@pending_bad_consequence.isEmpty))
      @pending_bad_consequence.substractVisibleTreasure(tesoro)
    end
    dieIfNoTreasures
  end
  
  
  def discardHiddenTreasure(tesoro)
    @hidden_treasures.delete(tesoro)
    if(@pending_bad_consequence!=nil and (!@pending_bad_consequence.isEmpty))
      @pending_bad_consequence.substractHiddenTreasure(tesoro)
    end
    dieIfNoTreasures
  end
  
  
  def validState()
    
    state = (@hidden_treasures.size < 5 and @pending_bad_consequence.isEmpty)
    return state
  end
  
  
  def initTreasures()
    dealer = CardDealer.instance
    dice = Dice.instance
    bringToLife
    treasure = dealer.nextTreasure
    @hidden_treasures << treasure
    number = dice.nextNumber
    if (number > 1)
      treasure = dealer.nextTreasure
      @hidden_treasures << treasure
    end
    if (number == 6)
      treasure = dealer.nextTreasure
      @hidden_treasures << treasure
    end
  end
  
 
  def stealTreasure()
    treasure = nil
    can_i = canISteal
    if (can_i)
      can_you = @enemy.canYouGiveMeATreasure
      if (can_you)
        treasure = @enemy.giveMeATreasure
        @hidden_treasures << treasure
        haveStolen
      end
    end
    return treasure
  end
  
  
  def setEnemy(enemy) #Player enemy
    @enemy = enemy
  end
  
  def canISteal()
    return @can_i_steal
  end
  
  def discardAllTreasures()
    
    tope = @visible_treasures.size   
    i=0
    while(i<tope)do
      discardVisibleTreasure(@visible_treasures.at(0))
      i=i+1
    end
    
    tope = @hidden_treasures.size    
    i=0
    while(i<tope)do
      discardHiddenTreasure(@hidden_treasures.at(0))
      i=i+1
    end
    
  end

  def getVisibleTreasures()
    return @visible_treasures
  end
  
  def getHiddenTreasures()
    return @hidden_treasures
  end
  
  def getEnemy()
    return @enemy
  end
  
  def canYouGiveMeATreasure()
    can_give = false
    can_give = true unless (@hidden_treasures.length < 1)
    return can_give                        
  end
  
  def giveMeATreasure()
    tesoro = @hidden_treasures[rand(@hidden_treasures.size)]
    @hidden_treasures.delete_at(@hidden_treasures.index(tesoro))
    return tesoro
  end
end

end

File: Napakalaki Java/NapakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author Germán
 */

import GUI.Dice;

public class Player {
    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead;
    private boolean canISteal;
    protected Player enemy;
    private BadConsequence pendingBadConsequence = null;
    private ArrayList<Treasure> VisibleTreasures = null;
    private ArrayList<Treasure> HiddenTreasures = null;
    
    
    //Metodos con visibilidad privada
    
    private void bringToLife(){
    
        dead = false;
    }
    
    private void incrementLevels(int i){
    
        level += i;
        if(level >10)
           level = 10;
        
    }
    
    private void decrementLevels(int i) {
    
        level -= i;
        if(level < 1)
            level = 1;
    }
    
    private void setPendingBadConsequence(BadConsequence b) {
        
        pendingBadConsequence = b;
    }
    
    private void applyPrize(Monster m) {
    
        int nLevels = m.getLevelsGained();
        incrementLevels(nLevels);
        int nTreasures = m.getTreasuresGained();
        Treasure treasure;
        CardDealer dealer;
        
        if(nTreasures>0){
            
            dealer = CardDealer.getInstance();
            for(int i=0;i<nTreasures;i++){
                treasure = dealer.nextTreasure();
                HiddenTreasures.add(treasure);
            }
            
        }
    }
    
    private void applyBadConsequence(Monster m) {
    
        BadConsequence badconsequence = m.GetMalRollo();
        int nLevels = badconsequence.getLevel();
        decrementLevels(nLevels);
        BadConsequence pendingBad = badconsequence.adjustToFitTreasureLists(VisibleTreasures, HiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure t) {
    
        boolean permitido = true;
        int contador =0;    
        int i;
        
        if(t.getType() == TreasureKind.ONEHAND){
            for(i=0;i<VisibleTreasures.size() && permitido;i++){
                if((VisibleTreasures.get(i)).getType() == TreasureKind.BOTHHANDS)
                    permitido = false;
            }
            
            if(permitido){
                for(i=0;i<VisibleTreasures.size();i++){
                    if((VisibleTreasures.get(i)).getType() == TreasureKind.ONEHAND)
                        contador++;
                }
            }
            
            if(contador == 2)
                permitido = false;
        }else{
           if(t.getType() == TreasureKind.BOTHHANDS){
               for(i=0;i<VisibleTreasures.size() && permitido;i++){
                if((VisibleTreasures.get(i)).getType() == TreasureKind.ONEHAND)
                    permitido = false;
                }
               
               for(i=0;i<VisibleTreasures.size() && permitido;i++){
                if((VisibleTreasures.get(i)).getType() == TreasureKind.BOTHHANDS)
                    permitido = false;
                }
           }else{
               for(i=0;i<VisibleTreasures.size();i++){
                    if(t.getType() == (VisibleTreasures.get(i)).getType())
                      permitido=false; 
                }    
           }
        }
           
        return permitido;
    }
    
    private int howManyVisibleTreasures (TreasureKind tkind) {
    
        int i;
        int contador=0;
    
        for(i=0; i < VisibleTreasures.size();i++){
            if((VisibleTreasures.get(i)).getType() == tkind)
                contador++;
        }
        
        return contador;
    }
    
    private void dieIfNoTreasures(){
    
        if(VisibleTreasures.isEmpty() && HiddenTreasures.isEmpty()){
            
            dead = true;
            level = 1;
            
        }
    }
    
    private void haveStolen() {
    
        canISteal = false;
    }
    
    //Metodos con visibilidad publica
    
    public Player(String nombre) {
        name = nombre;
        level = 1;
        dead = true;
        canISteal = true;
        enemy = null;
        pendingBadConsequence = new NumericBadConsequence("",0,0,0);
        VisibleTreasures = new ArrayList();
        HiddenTreasures = new ArrayList();
    }
    
    
    //Metodos redefinidos en la clase CultistPlayer-----------------
    //--------------------------------------------------------------
    protected Player (Player p) {
        this.name = p.getName();
        this.level = p.getLevels();
        this.dead = p.isDead();
        this.canISteal = p.canISteal();
        this.enemy = p.enemy;
        this.pendingBadConsequence = p.pendingBadConsequence;
        this.VisibleTreasures = p.getVisibleTreasures();
        this.HiddenTreasures = p.getHiddenTreasures();
    }
    
    protected Treasure giveMeATreasure () {
    
        Treasure tesoro;
        Random rand = new Random();
        int resultado = rand.nextInt(HiddenTreasures.size());
        tesoro = HiddenTreasures.get(resultado);
        HiddenTreasures.remove(tesoro);
        return tesoro;
    }
    
    public boolean canYouGiveMeATreasure() {
    
        boolean me_pueden_robar = false;
        
        if(HiddenTreasures.size() > 0)
                me_pueden_robar = true;
        
        return me_pueden_robar;
    }
    
    protected int getOponentLevel(Monster m) {
        return m.GetCombatLevel();
    }
    
    protected boolean shouldConvert() {
        Dice dado = Dice.getInstance();
        int resultado = dado.nextNumber();
        return (resultado == 1);
    }
    
    public int getCombatLevel() {
        
        int i,nivel;
        nivel = level;    
    
        for(i=0;i<VisibleTreasures.size();i++){
            nivel += (VisibleTreasures.get(i)).getBonus();
        }
        
        return nivel;  
        
    }
    //--------------------------------------------------------------
    
    @Override
    public String toString() {
        return name; 
    }
    
    public String getName() {
    
        return name;
    }
    
    public int getLevels() {
    
        return level;
    }
    
    public Player getEnemy() {
        return enemy;
    }
    
    public BadConsequence getPendingBC(){
        return pendingBadConsequence;
    }
    
    public boolean isDead() {
    
        return dead;
    }
    
    //Es necesario definir este metodo para poder usarlo en la clase CultistPlayer
    //en giveMeATreasure
    protected void setVisibleTreasures(ArrayList<Treasure> ts) {
        VisibleTreasures = ts;
    }
    
    public ArrayList<Treasure> getHiddenTreasures () {
    
        return HiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures () {
    
        return VisibleTreasures;
    }
    
    
    
    public CombatResult combat (Monster m) {
    
        CombatResult resultado;
        
        int myLevel = this.getCombatLevel();
        int monsterLevel = getOponentLevel(m);

        if(myLevel > monsterLevel){
        
            resultado = CombatResult.WIN;
            applyPrize(m);
        }else{
            
            resultado = CombatResult.LOSE;
            applyBadConsequence(m);    
        }
        
        if(resultado == CombatResult.LOSE) {
            if(this.shouldConvert())
                resultado = CombatResult.LOOSEANDCONVERT;
        }
        
        if(level > 9)
            resultado = CombatResult.WINGAME;
        
        return resultado;
    }
    
    public void makeTreasureVisible (Treasure t) {
    
        boolean canI = canMakeTreasureVisible(t);
        
        if(canI){
            VisibleTreasures.add(t);
            HiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t) {
    
        VisibleTreasures.remove(t);
        
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
        
            pendingBadConsequence.substractVisibleTreasure(t);
        }
        
        dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t) {
    
        HiddenTreasures.remove(t);
        
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
        
            pendingBadConsequence.substractHiddenTreasure(t);
        }
        
        dieIfNoTreasures();
        
    }
    
    public boolean validState() {
        boolean valido;
    
        valido = HiddenTreasures.size() < 5 && pendingBadConsequence.isEmpty();
        return valido;
    }
    
    public Treasure stealTreasure() {
    
        Treasure treasure = null;
        boolean canI = canISteal();
        boolean canYou;
        
        if(canI){
            
            canYou = enemy.canYouGiveMeATreasure();
            if(canYou){
                
                treasure = enemy.giveMeATreasure();
                HiddenTreasures.add(treasure);
                haveStolen();
            }
        }
        
        return treasure;
    }
    
    public void setEnemy(Player enemigo) {
    
        enemy = enemigo;
    }
    
    public boolean canISteal () {
    
        return canISteal;
    }
    
    public void discardAllTreasures() {
    
        while(!VisibleTreasures.isEmpty()){
            discardVisibleTreasure(VisibleTreasures.get(0));
        }  
        
        while (!HiddenTreasures.isEmpty()){
            discardHiddenTreasure(HiddenTreasures.get(0));
        }  
        
    }
    
    public void initTreasures(){
    
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        
        bringToLife();
        Treasure tesoro = dealer.nextTreasure();
        HiddenTreasures.add(tesoro);
        
        int number = dice.nextNumber();
        if(number>1){
            tesoro = dealer.nextTreasure();
            HiddenTreasures.add(tesoro);
        }
        
        if(number ==6){
            tesoro = dealer.nextTreasure();
            HiddenTreasures.add(tesoro);
        }
      
    }
 
}


//Para barajar: Collection.shuffle(mi_array)